xmrstak/misc/environment.hpp

```diff
@@ -12,35 +12,30 @@ class params;
 
 struct environment
 {
-
-	static environment& inst()
-	{
-		static environment env;
-		return env;
-	}
-	
-	environment& operator=(const environment& env)
+	static inline environment& inst(environment* init = nullptr)
 	{
-		this->pPrinter = env.pPrinter;
-		this->pglobalStates = env.pglobalStates;
-		this->pJconfConfig = env.pJconfConfig;
-		this->pExecutor = env.pExecutor;
-		this->pParams = env.pParams;
-		return *this;
-	}
+		static environment* env = nullptr;
 
+		if(env == nullptr)
+		{
+			if(init == nullptr)
+				env = new environment;
+			else
+				env = init;
+		}
 
-	environment() : pPrinter(nullptr), pglobalStates(nullptr)
-	{
+		return *env;
 	}
 
+	environment()
+	{
+	}
 
-	printer* pPrinter;
-	globalStates* pglobalStates;
-	jconf* pJconfConfig;
-	executor* pExecutor;
-	params* pParams;
-
+	printer* pPrinter = nullptr;
```

Comment:
@fireice-uk Could you please point me to a reference which describe why it is allowed to initialize member out of the constructor. Is this because of initializer lists?

Reply:
Ahh http://en.cppreference.com/w/cpp/language/data_members

Reply:
That's correct. For C++11 both those styles are equivalent. I chose the second one because it is clearer and nullptr is a constexpr.